### ISSUE TITLE
Lazy loading layers

### ISSUE THREAD

**Issue by DanWismer** [NCC-CNC/wheretowork#265]
Only load in raster layers to view on the map when the user turns them on. Cap user with the amount of viewable layers. Additionally, remove layer from browser cache if the user is at the cap and wants to view another layer

Currently, if hidden is no, then all raster layers are loaded in. It would be best if we only load in the raster layer when the user toggles it.

**Comment by DanWismer** [comment]
Putting this here as a note: https://book.javascript-for-r.com/widgets-adv.html#widgets-adv-performances

**Comment by jeffreyhanson** [comment]
Looks like an awesome book - I haven't seen that one before!

**Comment by DanWismer** [comment]
made some progress on this. Lazy Loading will be a game changer! I have rasters loading in on one of our biggest WTW projects and it renders smoothly IMO.


https://user-images.githubusercontent.com/29556279/219174506-5b51dea7-4772-48b9-b729-afd7c2f8cc29.mp4

**Comment by ricschuster** [comment]
Awesome! 
I'm really looking forward to testing this out when its ready.

**Comment by DanWismer** [comment]
Before I get to deep, I was wondering if @ricschuster and @jeffreyhanson had thoughts on this:

My idea is to add a new icon beside the eye (hide/show). The new icon would indicate and set the behavior of loading and unloading the layer into the browser (ie. `leaflet::addRasterImage` and `leaflet::clearGroup`). This way, the  hide/show icon still works as it always has. And if a layer is not loaded, then the eye icon will also be grey. This way, the user can load and offload layers as needed without us having to set a "layer count" threshold as each project will be different with how many layers can be on the basemap before performance dips.   

The flow would be when the app inits, layers that are set to visible and not hidden are also loaded. All layers not set to visible and not hidden will have a grey eye. Then, to display a new layer, user loads it (clicking the load icon), which then enables the visible to red so it can be toggled on and off. Then when there are too many layers and performance is bad, the user can offload a layer clicking the load icon, which will grey out the eye and remove the layer from the browser.

**Comment by DanWismer** [comment]
Also, I should comment that the hide/show eye only toggles the css `hidden` property of the layer. This is why when the visibility is off, performance is still poor because the layer is still in the browser. It isn't until `leaflet::clearGroup` is called which removes it and frees things up.

**Comment by jeffreyhanson** [comment]
Hmm, I like the idea of giving the user more control, but I worry this might be a bit too confusing. E.g., I feel like it should be pretty easy for the user to toggle layers on and off, without needing to understand concepts like caching or loading. What do you think about having some routine that tries to balance which layers are loaded or unloaded? E.g., it could have some logic like this:

1. If a layer is toggled on, there are >= 5 layers layers loaded, there is >= 1 layers toggled off,  then unload the toggled off layer which was toggled on the longest time ago (ie., accounting for layers being toggled on, then off, then on again).
2. If a layer is toggled on, there are >= 5 layers layers loaded, there are 0 layers toggled off, and there are > 10 layers loaded, then display a warning that having too many layers toggled on will degrade performance. 
3. If a layer is toggled off, there are >= 5 layers loaded, then unload the layer.

**Comment by DanWismer** [comment]
I agree, adding another icon with loading behavior could trip up a user. sorry I am having some trouble reading the logic you have.
 
What about this which takes pieces from what you have above:

if a layer is set to visible, it is loaded and cached when project is imported. All other layers are not visible and not loaded (hidden is always not loaded and not visible).

When a user toggles on a layer to visible, that layer is loaded and cached. The cache can hold up to *x* amount of layers (would be cool if there is a way to create a threshold based off the browser memory size). Once the cache threshold is hit, every new layer set to visible still gets loaded, but is not added to the cache (performance still keeps getting worse). In turn, every visible layer in the cache that gets set to invisible gets offloaded (`leaflet::clearGroup`, performance is better).

Just so we are on the same page, I am referring to cache as layers that are in the browser, so when the user toggles the visibility on and off, that behavior is instant. When layers are not in the cache, when a layer goes from not visible to visible, it loads (`addRasterImage`), which takes a few moments to show up.

**Comment by jeffreyhanson** [comment]
Yeah, that sounds good! I guess I was suggesting to have a slightly more advanced version that (i) had a specific process for deciding which invisible layer should get unloaded when the threshold is exceeded (e.g., if we keep track of when each loaded layer was made visible, we could unload the layer that was made visible the longest ago), and (ii) ensuring that the unloading process is triggerd when a user makes a layer invisible and thresholds is exceeded (i.e., not just triggered when threshold is exceeded and a layer is made visisble).  The (ii) is important so that if a user decides to make lots of layers visisble (so they are loaded and made visisble) and notices that things are slowing down, they can start making layers invisible and unloading layers to get better performance. How does that sound?

**Comment by DanWismer** [comment]
I like this! 

Do you think we should timestamp each layer as it is loaded? I have a been playing around with a new `loaded` field on the `Feature` class, but it is currently just `TRUE` or `FALSE`. I can try giving it a `Date/Time`

**Comment by DanWismer** [comment]
... I thought I could work directly with the `MapManager` class but realized that for `Themes` things get tricky where the `render_on_map` and `update_on_map` `Theme` methods get called and loop through each `feature` (which is nice because you can map all the layers in one shot). However, this is making it challenging to keep track of what is loaded for all themes/features within the project. For example, I can track when each `feature` within a `Theme` has been loaded. But I am not finding a good way for tracking all the features across `Themes` which is needed for determining if we have reached the threshold. Apologies if I am not making sense ... just trying to think through this some more

**Comment by DanWismer** [comment]
Here is the logic I have for mapping `Features`. If there are 4 loaded features, then offload a feature next time invisible is selected nothing fancy but a start
https://github.com/NCC-CNC/wheretowork/blob/21bf181624ac284fefa143dd8340c35d7a7cd721/R/class_Theme.R#L487-L516

**Comment by DanWismer** [comment]
stuck with getting lazy load to work with vector ....

For some reason when I remove the polygon, then render it again, it doesn't place the polygon into the new map pane. Rather it puts it back in the original pane. I don't get this issue with raster

I am wondering if this has something to do with it: https://github.com/Leaflet/Leaflet/issues/6408

https://itncc-my.sharepoint.com/:v:/g/personal/daniel_wismer_natureconservancy_ca/EZFUwYZJMl5PvCyb5er6I3QB0-w6CiKMZzwfiaS6hOun7g?e=YAdHvE

**Comment by DanWismer** [comment]
Had a nice break through today.

What I ended up doing was creating a new `pane` attribute on the `Feature` , `Weight` , `Include` , `Exclude` and `Solution` class. The `pane` attribute stores `paste(uuid::UUIDgenerate(), variable$index, sep = "-")` and is used as the html div class name for `leaflet::addMapPane(x, pane_id, zindex, visible)`. 

To clear layers, I ended up deleting the entire pane div: https://github.com/NCC-CNC/wheretowork/blob/d138a8f8709ee119ed17144ca39c6ae6a2333d77/R/utils_shinyjs.R#L114-L121

https://github.com/NCC-CNC/wheretowork/blob/d138a8f8709ee119ed17144ca39c6ae6a2333d77/R/class_MapManager.R#L483

And when a layer pane has been deleted and needs to be reloaded, I supply a new pane id.
https://github.com/NCC-CNC/wheretowork/blob/d138a8f8709ee119ed17144ca39c6ae6a2333d77/R/class_MapManager.R#L453
https://github.com/NCC-CNC/wheretowork/blob/d138a8f8709ee119ed17144ca39c6ae6a2333d77/R/class_Weight.R#L457 

This way we have a separation between the `app_data` ids which don't change after the app inits, And `pane` class names which are dynamically generated and can change as a deleted pane gets added back to the DOM. And only the layers that are loaded can be found in the DOM.
